### PR TITLE
[Formal] Expose Btor2 target

### PR DESCRIPTION
**Pull request by dobios**
This PR exposes the `btor2` target in Chisel. This is done through an `emitBtor2` method in `ChiselStage`.  
This currently depends on a [CIRCT PR to support `AssertProperty`](https://github.com/llvm/circt/pull/6974). 

### Contributor Checklist

- [x] Did you add Scaladoc to every public function/method?
- [x] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [ ] Did you add appropriate documentation in `docs/src`?
- [x] Did you request a desired merge strategy?
- [x] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement

- Feature (or new API)

#### Desired Merge Strategy

- Squash: The PR will be squashed and merged (choose this if you have no preference).

#### Release Notes  
 - [Formal] Allow for Chisel to be compiled to `btor2` for Bounded Model Checking using `ChiselStage.emitBtor2`.

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels? (Select the most appropriate one based on the "Type of Improvement")
- [ ] Did you mark the proper milestone (Bug fix: `3.6.x`, `5.x`, or `6.x` depending on impact, API modification or big change: `7.0`)?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you do one of the following when ready to merge:
  - [ ] Squash: You/ the contributor `Enable auto-merge (squash)`, clean up the commit message, and label with `Please Merge`.
  - [ ] Merge: Ensure that contributor has cleaned up their commit history, then merge with `Create a merge commit`.

### CODE REVIEW

File: src/test/scala/circtTests/stage/ChiselStageSpec.scala [hunk 1]
```diff
@@ -1095,6 +1095,22 @@ class ChiselStageSpec extends AnyFunSpec with Matchers with chiselTests.Utils {
...
+      class Counter extends Module {
+        val count = RegInit(0.U(32.W))
+        when(count === 42.U) { count := 0.U }
+        .otherwise { count := count + 1.U }
+        
+        AssertProperty((count <= 42.U))
+      }
+
+      val btor2 = ChiselStage.emitBtor2(new ChiselStageSpec.Foo)
+      println(btor2)
```
Comment (sequencer) [inline src/test/scala/circtTests/stage/ChiselStageSpec.scala:1111]: you can check the contents from btor2, but please don't print it out

**Review by sequencer — COMMENTED** [review]
Nice! can't wait to use this feature! My repo haven't go though the formal pass since I upgrade to chisel5.

**Comment by dobios** [comment]
Hi yes the goal is to check the contents, I'm just printing it for now for debugging purposes (which is why the PR is still a draft). On Apr 30, 2024 17:48, Jiuyang Liu ***@***.***> wrote:
@sequencer commented on this pull request.

Nice! can't wait to use this feature! My repo haven't go though the formal pass since I upgrade to chisel5.

In src/test/scala/circtTests/stage/ChiselStageSpec.scala:
> @@ -1095,6 +1095,22 @@ class ChiselStageSpec extends AnyFunSpec with Matchers with chiselTests.Utils {
       info(s"'$expectedOutput' exists")
     }
 
+    it("should emit btor2 to string") {
+      import chisel3._
+      import chisel3.ltl._
+
+      class Counter extends Module {
+        val count = RegInit(0.U(32.W))
+        when(count === 42.U) { count := 0.U }
+        .otherwise { count := count + 1.U }
+        
+        AssertProperty((count <= 42.U))
+      }
+
+      val btor2 = ChiselStage.emitBtor2(new ChiselStageSpec.Foo)
+      println(btor2)

you can check the contents from btor2, but please don't print it out

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>
[
{
***@***.***": "http://schema.org",
***@***.***": "EmailMessage",
"potentialAction": {
***@***.***": "ViewAction",
"target": "https://github.com/chipsalliance/chisel/pull/4035#pullrequestreview-2032863461",
"url": "https://github.com/chipsalliance/chisel/pull/4035#pullrequestreview-2032863461",
"name": "View Pull Request"
},
"description": "View this Pull Request on GitHub",
"publisher": {
***@***.***": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

File: src/test/scala/circtTests/stage/ChiselStageSpec.scala (hunk 1 above)
Comment (dobios) [reply src/test/scala/circtTests/stage/ChiselStageSpec.scala:1111]: Yes the goal is to check the contents, I'm just printing it for now for debugging purposes (which is why the PR is still a draft).

**Comment by dobios** [comment]
The [blocking PR](https://github.com/llvm/circt/pull/6974) has been merged in and should be included in the next firtool release, which would make this work.

**Comment by dobios** [comment]
The new [firtool release](https://github.com/llvm/circt/releases/tag/firtool-1.75.0) now supports everything that is needed for this PR!

**Review by jackkoenig — APPROVED** [review]
Looks great! Awesome work.

**Pull request merged**